hypervisor/arch/x86/sgx.c

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (C) 2019 Intel Corporation. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <types.h>
+#include <errno.h>
+#include <cpufeatures.h>
+#include <cpu_caps.h>
+#include <sgx.h>
+#include <cpuid.h>
+#include <vm.h>
+#include <logmsg.h>
+
+#define SGX_OPTED_IN (MSR_IA32_FEATURE_CONTROL_SGX_GE | MSR_IA32_FEATURE_CONTROL_LOCK)
+
+/* For the static variables, which are not explicitly initialzed will be inited to 0 */
+static int32_t init_sgx_ret = 0;
+static struct epc_section pepc_sections[MAX_EPC_SECTIONS];			/* physcial epc sections */
+static struct epc_map vm_epc_maps[MAX_EPC_SECTIONS][CONFIG_MAX_VM_NUM];		/* epc resource mapping for VMs */
+
+/* Partition physcial EPC resource according to VM configurations.
+ * Build the mappings between HPA and GPA for EPT mapping later.
+ * EPC resource partition and mapping relationship will stay the same after sgx init.
+ */
+static int32_t partition_epc(void)
+{
+	uint32_t eax, ebx, ecx, edx, type;
+	uint32_t psec_id, nvm_id, mid;
+	uint64_t psec_addr = 0UL, psec_size = 0UL;
+	uint64_t vm_request_size = 0UL, free_size = 0UL, alloc_size;
+	struct acrn_vm_config *vm_config;
+	int32_t ret = 0;
+
+	for (psec_id = 0U, nvm_id = 0U; psec_id < MAX_EPC_SECTIONS && nvm_id < CONFIG_MAX_VM_NUM; ) {
+		if (vm_request_size ==0U) {
+			mid = 0U;
+			vm_config = get_vm_config(nvm_id);
+			if((((vm_config->epc.size | vm_config->epc.base) & ~PAGE_MASK) != 0UL) ||
```

Comment:
Strange..    (size | base） & 0xfff != 0UL..   This comes from HV code (configuration), we don't need to check if we have mistake in configuration.  Add comments to size and base for PAGE ALIGNMENT is good enough

Reply:
if this is for santizing, then do it in sanitize_vm_config

Reply:
I put the checking inside SGX because the feature is kind of independent and if SGX is not needed, we can get rid of the related code as a whole, what do you think?

Reply:
ok